hudi-spark-datasource/hudi-spark-common/src/main/scala/org/apache/spark/sql/vectorized/ColumnarBatchUtils.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.spark.sql.vectorized;
+
+import org.apache.hudi.common.util.collection.Pair;
+
+import org.apache.spark.sql.types.StructType;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+
+public class ColumnarBatchUtils {
+
+  public static UnaryOperator<ColumnarBatch> generateProjection(StructType from, StructType to) {
+    if (from.length() < to.length()) {
+      throw new IllegalStateException(from + " has less columns than " + to);
+    }
+
+    if (from.equals(to)) {
+      return UnaryOperator.identity();
+    }
+
+    int[] projection = new int[to.size()];
+    for (int i = 0; i < to.length(); i++) {
+      projection[i] = from.fieldIndex(to.fields()[i].name());
+    }
+
+    return columnarBatch -> {
+      ColumnVector[] vectors = new ColumnVector[projection.length];
+      for (int i = 0; i < projection.length; i++) {
+        vectors[i] = columnarBatch.column(projection[i]);
+      }
+
+      //do this until we get rid of spark 2
+      ColumnarBatch b = new ColumnarBatch(vectors);
+      b.setNumRows(columnarBatch.numRows());
+      return b;
+    };
+  }
+
+  public static UnaryOperator<ColumnarBatch> generateProjectionInplace(StructType from, StructType to) {
```

Comment:
I don't have it using the in place projection, but we can do it if there are memory issues

Reply:
We can actually extend columnar batch after we get rid of spark<3.3 to do it inplace without the swapping